### ISSUE TITLE
Make algorithm JS only

### ISSUE THREAD

**Issue by dest4** [adblockradio/adblockradio#4]
Communication between Node and Python is a pain for portability. Furthermore, the native zeromq IPC lib has to be built at installation. Damn, let's make Adblock Radio algorithm JS only, so that it could be shipped as a big Node.js file (and, with a proper solution for ffmpeg, as a browser JS in the future).

- [x]  Tensorflow has a JS implementation https://js.tensorflow.org/setup/
- [x] Keras models can be converted to Tensorflow JS format. https://js.tensorflow.org/tutorials/import-keras.html
- [x] Computation of Mel Frequency Cepstra Coefficients (MFCC): that library seems to do the job https://github.com/vail-systems/node-mfcc. But the Python implementation seems more complex, with for example preemphasis filter https://github.com/jameslyons/python_speech_features/blob/master/python_speech_features/base.py#L66 => do we need to reimplement that in JS?

**Comment by dest4** [comment]
I have written a MFCC calculation in JS. Still having problems with Tensorflow.JS though, but it's making progress.

**Comment by dest4** [comment]
New branch showing work in progress:
https://github.com/adblockradio/adblockradio/tree/mljs

Still need to validate the calculations.

**Comment by dest4** [comment]
MFCC calculations are now validated. Testing the new algo end to end now.

**Comment by dest4** [comment]
End to end validated on the radio recording (in `podcasts/example.mp3`), yay!
About 8 minutes of content analyzed on a Core I5-5200U in
- 58s with the Node + Python version
- 2min21 with the pure Node version, with native Tensorflow library.

It's a huge discrepancy, but I noticed that the Python calculus is massively parallel while the Node version uses only one core and probably one thread. The test CPU has two cores and 4-theads.
In conclusion, the two approaches seem equivalently efficient.

**Comment by dest4** [comment]
Here is diff of the results log. Differences are only due to rounding errors, most often about the gain.

```
$ diff example.mp3.keras.json example.mp3.tfjs.json 
2102c2102
< 			"gain": 76.47,
---
> 			"gain": 76.48,
2152c2152
< 			"gain": 76.58,
---
> 			"gain": 76.59,
3056c3056
< 			"gain": 77.17,
---
> 			"gain": 77.18,
3446c3446
< 			"gain": 75.87,
---
> 			"gain": 75.88,
3762c3762
< 			"gain": 74.21,
---
> 			"gain": 74.22,
4229c4229
< 					0.253,
---
> 					0.254,
4260c4260
< 				0.2515,
---
> 				0.252,
4268c4268
< 			"gain": 75.61,
---
> 			"gain": 75.62,
6086c6086
< 			"gain": 78.66,
---
> 			"gain": 78.67,
6376c6376
< 			"gain": 77.96,
---
> 			"gain": 77.97,
7778c7778
< 			"gain": 77.06,
---
> 			"gain": 77.07,
8326c8326
< 			"gain": 75.76,
---
> 			"gain": 75.77,
9662c9662
< 			"gain": 75.6,
---
> 			"gain": 75.61,
10052c10052
< 			"gain": 75.16,
---
> 			"gain": 75.17,
11646c11646
< 			"gain": 74.68,
---
> 			"gain": 74.69,
12020c12020
< 			"gain": 76.3,
---
> 			"gain": 76.31,
12766c12766
< 			"gain": 76.15,
---
> 			"gain": 76.16,
13048c13048
< 			"gain": 76.26,
---
> 			"gain": 76.27,
13570c13570
< 			"gain": 75.49,
---
> 			"gain": 75.5,
14110c14110
< 			"gain": 76.98,
---
> 			"gain": 76.99,
15156c15156
< 			"gain": 75.76,
---
> 			"gain": 75.77,
16974c16974
< 			"gain": 76.2,
---
> 			"gain": 76.21,
17306c17306
< 			"gain": 75.58,
---
> 			"gain": 75.59,
17480c17480
< 			"gain": 76.53,
---
> 			"gain": 76.54,
18078c18078
< 			"gain": 74.76,
---
> 			"gain": 74.77,
18302c18302
< 			"gain": 74.47,
---
> 			"gain": 74.48,
19414c19414
< 			"gain": 74.52,
---
> 			"gain": 74.53,
21024c21024
< 			"gain": 75.5,
---
> 			"gain": 75.51,
21630c21630
< 			"gain": 76.68,
---
> 			"gain": 76.69,
22650c22650
< 			"gain": 76.54,
---
> 			"gain": 76.55,
22708c22708
< 			"gain": 72.23,
---
> 			"gain": 72.24,
23538c23538
< 			"gain": 76.83,
---
> 			"gain": 76.84,
23596c23596
< 			"gain": 74.27,
---
> 			"gain": 74.28,
24110c24110
< 			"gain": 74.58,
---
> 			"gain": 74.59,
24516c24516
< 			"gain": 76.05,
---
> 			"gain": 76.06,
25428c25428
< 			"gain": 75.01,
---
> 			"gain": 75.02,
26521c26521
< 	"analysisTime": 58.593
---
> 	"analysisTime": 141.5
```

**Comment by dest4** [comment]
On low horse power CPUs, the single thread approach does not work well. I may have to put the computations in a separate thread.

**Comment by dest4** [comment]
I have put the ML computations in a different thread in https://github.com/adblockradio/adblockradio/commit/e5f1340675f7727982d7979c15ae255a67510da1

Now the podcast analysis demo, which did run in 2min21 in the single thread version, runs in 2min.

I had reviewed [node-webworker-threads](https://github.com/audreyt/node-webworker-threads) and would have loved to provide a unified codebase for node and browser, but it is not possible to `require` in webworkers. Another solution will be needed for inference in browsers.

**Comment by dest4** [comment]
I have merged PR https://github.com/adblockradio/adblockradio/pull/8

Now computations can be done with Python (default) or with JS+native Tensorflow lib, with pure JS Tensorflow as a fallback.

The JS+native lib performance is underwhelming unfortunately, but it's a nice perspective to have it in order to port Adblock Radio to web browser extensions some day.
